### ISSUE TITLE
Regarding Indexing Benchmark

### ISSUE THREAD

**Issue by netw0rkf10w** [facebookresearch/xformers#780]
# ❓ Questions and Help

Hello,

Thank you very much for open sourcing your indexing ops! Would it be possible to share the benchmarking results please? 
(I'm referring to https://github.com/facebookresearch/xformers/blob/main/xformers/benchmarks/benchmark_indexing.py)

In particular I would be interested in knowing how much faster your custom ops are compared to the equivalent PyTorch implementation.

Thanks a lot for your response!

**Comment by netw0rkf10w** [comment]
(Forgot to tag @danthe3rd who was the one to publish that code)

**Comment by danthe3rd** [comment]
Hi,

This is the result of running the benchmarks script on an A100 GPU:
```
[--------------------------- scaled_index_addfw ---------------------------]                                                                                                                                
                                                    |  optimized  |  pytorch
1 threads: -----------------------------------------------------------------
      f16 B_src=28, B_out=48, M=1, D=394752 s=N     |      96.7   |   279.9 
      f16 B_src=28, B_out=48, M=1, D=394752 s=Y     |    1023.2   |   329.8 
      f16 B_src=28, B_out=48, M=257, D=1536 s=N     |      98.1   |   280.2 
      f16 B_src=28, B_out=48, M=257, D=1536 s=Y     |      99.0   |   329.8 
      f16 B_src=115, B_out=192, M=50, D=1536 s=N    |      77.3   |   222.2 
      f16 B_src=7401, B_out=12336, M=1, D=1536 s=N  |      97.7   |   299.5 
      f16 B_src=5760, B_out=9600, M=1, D=1536 s=N   |      77.6   |   236.6 
      f16 B_src=2467, B_out=12336, M=1, D=1536 s=N  |      69.8   |   136.2 

Times are in microseconds (us).

[-------------------------- scaled_index_addfwbw --------------------------]                                                                                                                                
                                                    |  optimized  |  pytorch
1 threads: -----------------------------------------------------------------
      f16 B_src=28, B_out=48, M=1, D=394752 s=N     |     920.3   |   473.1 
      f16 B_src=28, B_out=48, M=1, D=394752 s=Y     |    2124.9   |   523.1 
      f16 B_src=28, B_out=48, M=257, D=1536 s=N     |     179.7   |   473.5 
      f16 B_src=28, B_out=48, M=257, D=1536 s=Y     |     241.8   |   522.9 
      f16 B_src=115, B_out=192, M=50, D=1536 s=N    |     176.3   |   375.8 
      f16 B_src=7401, B_out=12336, M=1, D=1536 s=N  |     182.4   |   520.6 
      f16 B_src=5760, B_out=9600, M=1, D=1536 s=N   |     177.3   |   409.5 
      f16 B_src=2467, B_out=12336, M=1, D=1536 s=N  |     176.1   |   242.8 

Times are in microseconds (us).

[-------------------------------- index_selectfw --------------------------------]                                                                                                                          
                                                          |  optimized  |  pytorch
1 threads: -----------------------------------------------------------------------
      f16 D=1536 batches=((48, 257), (50, 192)) keep=0.6  |     76.8    |   227.7 

Times are in microseconds (us).

[------------------------------- index_selectfwbw -------------------------------]                                                                                                                          
                                                          |  optimized  |  pytorch
1 threads: -----------------------------------------------------------------------
      f16 D=1536 batches=((48, 257), (50, 192)) keep=0.6  |    487.5    |   882.5 

Times are in microseconds (us).
```

These operators are not optimized for all shapes, so please benchmark your specific use-case before, as they might be significatively slower than they pytorch counterparts sometimes.
The trick to be faster than PyTorch here is that we assume the indices have no duplicate, which removes a lot of complexity (the PyTorch kernels have to use atomic adds etc...)

**Comment by netw0rkf10w** [comment]
@danthe3rd That's awesome! Thanks a lot for sharing! The speedups are very impressive. 

By the way, you mentioned in the [changelog](https://github.com/facebookresearch/xformers/blob/main/CHANGELOG.md#added-3) that these functions "can be used to write efficient stochastic depth in transformer architectures". I have seen such implementation in the code of DINOv2, and it seems to me that these ops are only useful for the variant of stochastic depth in which we randomly select subsets of the current batch (if we instead skip the entire batch as in the original stochastic depth, then these ops have no use). Is that correct? (Sorry if it's not appropriate to ask this question here, I was thinking about asking in the DINOv2 repo but I saw that you're also a contributor there and because of the changing, I feel that you know the answer. Thanks!)

**Comment by danthe3rd** [comment]
> I have seen such implementation in the code of DINOv2

Yes that's correct - we have released these operators at the same time as the DINOv2 codebase.

> if we instead skip the entire batch as in the original stochastic depth, then these ops have no use. Is that correct?

Yes this is correct. Skipping the entire batch was not very convenient in our training setup (Fully Sharded Data Parallel) - as we would need to skip it for all GPUs, otherwise some nodes would not have any gradient for some layers.

Cc @qasfb who might have more context on that

**Comment by netw0rkf10w** [comment]
@danthe3rd Thanks for the confirmation! 

> Skipping the entire batch was not very convenient in our training setup (Fully Sharded Data Parallel) - as we would need to skip it for all GPUs, otherwise some nodes would not have any gradient for some layers.

I'm not familiar with FSDP training, but I guess the issue you mentioned is similar to why we have to set `find_unused_parameters=True` in DDP when skipping the entire batch. I proposed a solution to this problem in the PyTorch repo (https://github.com/pytorch/pytorch/issues/90171) but it doesn't seem to interest the team (if this solution is implemented, which I'm unable to do, then I believe that I'll have an even more efficient implementation of stochastic depth).
